### PR TITLE
CY-1471 - Added CircleCI npm publish configuration

### PR DESCRIPTION
**Pull request by qooban**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -33,8 +35,32 @@ jobs:
...
+          filters:
+            tags:
+              only: /.*/
+      - deploy:
+          requires:
+            - build-test
+          filters:
+            tags:
+              only: /^v.*/
+            branches:
```
Comment (tyacbovi) [inline .circleci/config.yml]: I think this is what we want per-version release and dev release, no?
            branches:
              only:
                - master

File: .circleci/config.yml [hunk 2]
```diff
@@ -33,8 +35,32 @@ jobs:
...
+          command: npm publish
+
 workflows:
   version: 2
-  build-test:
+  build-test-deploy:
     jobs:
-      - build-and-test
+      - build-test:
+          filters:
```
Comment (tyacbovi) [inline .circleci/config.yml]: Why do we need this?

Comment (qooban) [reply .circleci/config.yml]: CIrcleCI reasons: https://circleci.com/docs/2.0/workflows/#git-tag-job-execution

File: README.md [hunk 3]
```diff
@@ -62,25 +62,31 @@ Some general guidelines for different type of assets are listed below.
...
 ## Deployment
 
-Cloudify UI Common library is deployed to [NPM](https://www.npmjs.com). There are mainly two supported options how to deploy new version. Both are described below.
+Cloudify UI Common library is published in [NPM](https://www.npmjs.com) registry.
 
-### Manual
+Assuming all your changes are merged and you have up-to-date local `master` branch:
+1. Create new version using npm-version internal command, eg. 
+   
+   `npm version patch` for new patch version,
```
Comment (tyacbovi) [inline README.md]: In which case we will change patch and which minor?

File: ci/publish-version.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+echo "Creating cloudify-ui-common ${NEXT_VERSION}...";
+
+echo "Creating version publish branch...";
+NEXT_VERSION_BRANCH="publish-version-$NEXT_VERSION"
+git checkout -b ${NEXT_VERSION_BRANCH}
+git add package.json package-lock.json
+git commit -m "Bump version"
+git tag $NEXT_VERSION
+
+echo
```
Comment (tyacbovi) [inline ci/publish-version.sh]: You can add \n in the following echo

File: README.md [hunk 5]
```diff
@@ -62,25 +62,31 @@ Some general guidelines for different type of assets are listed below.
...
+1. Create new version using npm-version internal command, eg. 
+   
+   `npm version patch` for new patch version, 
+   
+   `npm version minor` for new minor version, 
+   
+   `npm version --help` for usage details.
 
-TODO: Use [np](https://github.com/sindresorhus/np) tool.
+1. Create new branch and push it to remote.
```
Comment (tyacbovi) [inline README.md]: With the name with the format?

File: ci/publish-version.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+echo "This is the last step before CircleCI-based npm publish.";
+echo "Are you sure you want to trigger cloudify-ui-common ${NEXT_VERSION} version publish?";
+select choice in "Yes" "No"; do
+case "$choice" in
+    Yes)
+        echo "Pushing changes";
+        git push --quiet --follow-tags origin ${NEXT_VERSION_BRANCH}
```
Comment (tyacbovi) [inline ci/publish-version.sh]: I think it's better without the quiet in case of issues

File: ci/publish-version.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+NEXT_VERSION_BRANCH="publish-version-$NEXT_VERSION"
+git checkout -b ${NEXT_VERSION_BRANCH}
+git add package.json package-lock.json
+git commit -m "Bump version"
+git tag $NEXT_VERSION
+
+echo
+echo "This is the last step before CircleCI-based npm publish.";
+echo "Are you sure you want to trigger cloudify-ui-common ${NEXT_VERSION} version publish?";
+select choice in "Yes" "No"; do
```
Comment (tyacbovi) [inline ci/publish-version.sh]: Why not prompt this in the beginning?

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cloudify-ui-common",
-  "version": "1.0.0-alpha.1",
+  "version": "1.0.0-alpha.10",
```
Comment (tyacbovi) [inline package.json]: That was fast ;)

File: package.json [hunk 9]
```diff
@@ -30,11 +30,15 @@
     "lint": "npm run lint:code",
     "lint:code": "eslint src test",
     "lint:docs": "documentation lint src/index.js",
-    "prepare": "npm test && npm run build",
+    "publish:dev": "npm run test:build && sh ci/publish-version.sh prerelease --preid=alpha",
```
Comment (tyacbovi) [inline package.json]: I don't if we need prerelease capability

File: .circleci/config.yml [hunk 10]
```diff
@@ -33,8 +35,56 @@ jobs:
...
+            git config credential.helper 'cache --timeout=120'
+            git config user.email "cosmo-admin@github.com"
+            git config user.name "cosmo-admin-CircleCI"
+
+      - run:
+          name: Update main branch
+          command: |
+            git checkout ${master_branch}
+            git merge --ff-only ${CIRCLE_BRANCH}
+            git push -q https://${gh_token_qooban}@github.com/cloudify-cosmo/cloudify-ui-common.git ${master_branch}
```
Comment (tyacbovi) [inline .circleci/config.yml]: gh_token_qooban I don't this is a good idea

File: README.md (hunk 3 above)
Comment (qooban) [reply README.md]: As I wrote in slack message, I did not update README.md, yet. Will do that today.

File: .circleci/config.yml [hunk 11]
```diff
@@ -33,8 +35,56 @@ jobs:
...
+            tags:
+              only: /.*/
+      - deploy:
+          requires:
+            - build-test
+          filters:
+            tags:
+              only: /^v.*/
+            branches:
+              only: /^publish-version-v.*/
```
[the remark this answers is not in the document]
Comment (qooban) [reply .circleci/config.yml]: Then sth like `visibility-code-fix` will also be taken. 
I know we have tags filter above, but I prefer to have something very specific nobody uses explicitly.

File: ci/publish-version.sh (hunk 4 above)
Comment (qooban) [reply ci/publish-version.sh]: Sure, can do.

File: README.md (hunk 5 above)
Comment (qooban) [reply README.md]: The same as https://github.com/cloudify-cosmo/cloudify-ui-common/pull/8#discussion_r307115626 .

File: ci/publish-version.sh (hunk 6 above)
Comment (qooban) [reply ci/publish-version.sh]: OK, can change.

File: ci/publish-version.sh (hunk 7 above)
Comment (qooban) [reply ci/publish-version.sh]: All the steps till that point were done locally, that's why there is no reason to be so careful. 
This switch is just before `git push`.
If you choose "No", then all the local changes will be reverted.

File: package.json (hunk 8 above)
Comment (qooban) [reply package.json]: I was testing on other branch ;-)

File: package.json (hunk 9 above)
Comment (qooban) [reply package.json]: Yeah, maybe it's not necessary.

File: .circleci/config.yml (hunk 10 above)
Comment (qooban) [reply .circleci/config.yml]: You mean that it's my token or GitHub token in general?

I talked with Yuval yesterday. After we finish testing he'll switch the tokens to the one from Cloudify-wide account and it'll be possible to merge to master.

For now I used mine and `master_branch` was set to my own branch (not master). Token's value is stored in CircleCI environmental variables and cannot be retrieved directly.

**Pull request merged**